store/store.go

```diff
@@ -785,39 +799,53 @@ func (s *Store[H]) nextTail(ctx context.Context) (tail H, changed bool) {
 	}
 }
 
-func (s *Store[H]) loadHeadAndTail(ctx context.Context) error {
+// init loads the head and tail headers and sets them on the store.
+// allows partial initialization of either tail or head if one of the is not found.
+func (s *Store[H]) init(ctx context.Context) error {
 	head, err := s.readByKey(ctx, headKey)
-	if err != nil {
-		return fmt.Errorf("header/store: cannot load headKey: %w", err)
+	if err != nil && !errors.Is(err, header.ErrNotFound) {
+		return fmt.Errorf("reading headKey: %w", err)
+	}
+	if !head.IsZero() {
+		s.contiguousHead.Store(&head)
+		s.heightSub.Init(head.Height())
+		log.Debugw("initialized head", "height", head.Height())
 	}
 
 	tail, err := s.readByKey(ctx, tailKey)
-	if err != nil {
-		return fmt.Errorf("header/store: cannot load tailKey: %w", err)
+	if err != nil && !errors.Is(err, header.ErrNotFound) {
+		return fmt.Errorf("reading tailKey: %w", err)
+	}
+	if !tail.IsZero() {
+		s.tailHeader.Store(&tail)
+		log.Debugw("initialized tail", "height", tail.Height())
 	}
 
-	s.init(head, tail)
 	return nil
 }
 
+// ensureInit initializes the store with the given headers if it is not already initialized.
 func (s *Store[H]) ensureInit(headers []H) {
-	headExist, tailExist := s.contiguousHead.Load() != nil, s.tailHeader.Load() != nil
-	if len(headers) == 0 || (tailExist && headExist) {
+	if len(headers) == 0 {
 		return
-	} else if tailExist || headExist {
-		panic("header/store: head and tail must be both present or absent")
 	}
 
-	tail, head := headers[0], headers[len(headers)-1]
-	s.init(head, tail)
-}
+	if headPtr := s.contiguousHead.Load(); headPtr == nil {
+		head := headers[len(headers)-1]
+		if s.contiguousHead.CompareAndSwap(headPtr, &head) {
+			s.heightSub.Init(head.Height())
+			log.Debugw("initialized head", "height", head.Height())
+		}
+	}
 
-func (s *Store[H]) init(head, tail H) {
-	s.contiguousHead.Store(&head)
-	s.heightSub.Init(head.Height())
-	s.tailHeader.Store(&tail)
+	if tailPtr := s.tailHeader.Load(); tailPtr == nil {
```

Comment:
Is there a case potentially where headers[0].height can actually be < tailHeader.Load.Height ?

Reply:
kk